### PR TITLE
Refactored client exceptions and added logic for job dir clean up after job finishes.

### PR DESCRIPTION
**Pull request by amitsharma1981**
- The job dependencies are now deleted (configurable) before the archive is created. The deletion had a bug which is now fixed by deleting each application dependencies one-by-one

- The clients now throw IllegalArgumentException in case of missing params

- The wait method in JobClient now does not gobble up the the InterruptedException from the thread sleep method but lets it propagate up to the caller to handle it.

- All non 2xx response in the clients now throw GenieClientExceptions and IOExceptions are thrown for Network and other IO problems.

### CODE REVIEW

File: genie-client/src/main/java/com/netflix/genie/client/JobClient.java [hunk 1]
```diff
@@ -440,14 +462,15 @@ public void killJob(final String jobId) throws IOException {
      *                     GenieClientException will be thrown
      * @param pollTime     the time to sleep between polling for job status
      * @return The job status for the job after completion
-     * @throws GenieClientException       For timeout or other errors.
-     * @throws InterruptedException on thread errors
-     * @throws IOException If the response received is not 2xx.
+     *
+     * @throws InterruptedException on thread errors.
+     * @throws GenieClientException If the response received is not 2xx.
```
Comment (skwslide) [inline genie-client/src/main/java/com/netflix/genie/client/JobClient.java]: The description of this exception is not accurate.

File: genie-client/src/main/java/com/netflix/genie/client/config/GenieNetworkConfiguration.java [hunk 2]
```diff
@@ -30,14 +30,19 @@
 @Setter
 public class GenieNetworkConfiguration {
 
+    /**
+     * Default network timeout value if not specified.
+     */
+    public static final long DEFAULT_TIMEOUT = -1;
```
Comment (skwslide) [inline genie-client/src/main/java/com/netflix/genie/client/config/GenieNetworkConfiguration.java]: Is -1 same as infinite? Infinite is bad (usually).

File: genie-client/src/main/java/com/netflix/genie/client/interceptor/ResponseMappingInterceptor.java [hunk 3]
```diff
@@ -52,8 +53,13 @@ public Response intercept(final Chain chain) throws IOException {
         if (response.isSuccessful()) {
             return response;
         } else {
-            final JsonNode responseBody = mapper.readTree(response.body().string());
-            throw new GenieClientException(response.code(), response.message() + " : " + responseBody.get("message"));
+            try {
+                final JsonNode responseBody = mapper.readTree(response.body().string());
+                throw new GenieClientException(response.code(), response.message()
+                    + " : " + responseBody.get("message"));
+            } catch (JsonMappingException jme) {
```
Comment (skwslide) [inline genie-client/src/main/java/com/netflix/genie/client/interceptor/ResponseMappingInterceptor.java]: Catching JsonProcessingException here will encompass JsonMappingException and JsonParseException. Or catch both explicitly.

File: genie-client/src/main/java/com/netflix/genie/client/JobClient.java [hunk 4]
```diff
@@ -461,34 +484,32 @@ public JobStatus waitForCompletion(final String jobId, final long blockTimeout,
...
-                    Thread.sleep(pollTime);
-                } else {
-                    throw new GenieClientException("Timed out waiting for job to finish");
-                }
-            } catch (InterruptedException ie) {
-                throw new IOException("Received interreupted exception from wait thread.");
+            if (System.currentTimeMillis() - startTime < blockTimeout) {
+                Thread.sleep(pollTime);
+            } else {
+                throw new GenieClientException("Timed out waiting for job to finish");
```
Comment (skwslide) [inline genie-client/src/main/java/com/netflix/genie/client/JobClient.java]: As already discussed, something like GenieTimeoutException will be better.

File: genie-client/src/main/java/com/netflix/genie/client/config/GenieNetworkConfiguration.java (hunk 2 above)
Comment (amitsharma1981) [inline genie-client/src/main/java/com/netflix/genie/client/config/GenieNetworkConfiguration.java]: -1 basically means its not set by user so i will let okhttp use defaults. which is believe is 10000.

Comment (skwslide) [inline genie-client/src/main/java/com/netflix/genie/client/config/GenieNetworkConfiguration.java]: Got it.

File: genie-client/src/main/java/com/netflix/genie/client/interceptor/ResponseMappingInterceptor.java [hunk 5]
```diff
@@ -57,7 +57,7 @@ public Response intercept(final Chain chain) throws IOException {
                 final JsonNode responseBody = mapper.readTree(response.body().string());
                 throw new GenieClientException(response.code(), response.message()
                     + " : " + responseBody.get("message"));
-            } catch (JsonMappingException jme) {
+            } catch (JsonProcessingException jpe) {
                 throw new GenieClientException(response.code(), response.message() + response.body().toString());
```
Comment (skwslide) [inline genie-client/src/main/java/com/netflix/genie/client/interceptor/ResponseMappingInterceptor.java]: I suspect calling response.body() a second time returns nothing.

Comment (amitsharma1981) [inline genie-client/src/main/java/com/netflix/genie/client/interceptor/ResponseMappingInterceptor.java]: Yea thats correct. I am going to cache it for reuse.

File: genie-client/src/main/java/com/netflix/genie/client/JobClient.java [hunk 6]
```diff
@@ -461,34 +485,33 @@ public JobStatus waitForCompletion(final String jobId, final long blockTimeout,
...
-     * @param id           the Genie job ID to wait for completion.
+     * @param jobId           the Genie job ID to wait for completion.
      * @param blockTimeout the time to block for (in ms), after which a
      *                     GenieClientException will be thrown.
      * @return The job status for the job after completion.
-     * @throws GenieClientException       For timeout or other errors.
-     * @throws InterruptedException on timeout/thread errors.
-     * @throws IOException If the response recieved is not 2xx.
+     *
+     * @throws InterruptedException on thread errors.
```
Comment (kilink) [inline genie-client/src/main/java/com/netflix/genie/client/JobClient.java]: Nitpick: I would probably reword this, as an InterruptedException does not necessarily indicate a "thread error". Here's what the Javadoc for Future.get says: "if the current thread was interrupted while waiting", which seems like a more accurate description.

I think the same Javadoc comment for InterruptedException is repeated elsewhere, so if we decide to change it, we should fix it everywhere.

**Comment by kilink** [comment]
LGTM in terms of addressing the issues I opened in #275 and #276. Thanks!

**Pull request merged**